dialogue-core/src/main/java/com/palantir/dialogue/core/HostAndLimitedChannels.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+import com.google.common.collect.ImmutableList;
+import com.palantir.logsafe.Preconditions;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import java.util.stream.IntStream;
+
+final class HostAndLimitedChannels {
+
+    private final ImmutableList<HostAndLimitedChannel> channels;
+    private final BiMap<HostIdx, HostAndLimitedChannel> lookups;
+
+    private HostAndLimitedChannels(ImmutableList<HostAndLimitedChannel> channels) {
+        this.channels = channels;
+        this.lookups = HashBiMap.create(channels.size());
+        channels.forEach(hostLimitedChannel -> lookups.put(hostLimitedChannel.getHostIdx(), hostLimitedChannel));
+    }
+
+    ImmutableList<HostAndLimitedChannel> getUnorderedChannels() {
+        return channels;
+    }
+
+    HostAndLimitedChannel getByUnordered(int index) {
+        return channels.get(index);
+    }
+
+    HostAndLimitedChannel getByHostIdx(HostIdx hostIdx) {
+        return Preconditions.checkNotNull(lookups.get(hostIdx), "Unknown hostIdx");
+    }
+
+    int unorderedIndexOf(HostAndLimitedChannel hostAndLimitedChannel) {
+        return channels.indexOf(hostAndLimitedChannel);
+    }
```

Comment:
Could use some javadoc describing the difference between unordered index and hostIndex index. It's hard to reason through why we have both.

Reply:
yeah, I wasn't quite sure how to best handle this, but in my original attempt the pin until error channel still kept it's own PinChannel thing, which was exactly the same as this thing, but allowed you to shuffle and refer to channels with just an List index. I'm open to suggestions on how to reword/restructure this, but basically the contract is that HostAndLimitedChannels maintains the original HostIdx -> LimitedChannel, even though you can scramble the ordering if you need it for the usecase (pin until error).

Reply:
Ok, I added a bit of javadoc, and removed the specilized weird methods. Instead just have getChannels and all the usual ImmutableList methods can be used directly.